compiler/optimizations/inferConstRefs.cpp

```diff
@@ -0,0 +1,377 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "optimizations.h"
+
+#include "astutil.h"
+#include "expr.h"
+#include "stmt.h"
+#include "stlUtil.h"
+#include <deque>
+
+class RefInfo {
+  public:
+    bool                 finalized; // Indicates if analysis is done
+    Symbol*              sym;
+    std::deque<SymExpr*> todo;
+    RefInfo(Symbol*);
+};
+
+RefInfo::RefInfo(Symbol* s) {
+  sym = s;
+  finalized = sym->qualType().getQual() == QUAL_CONST_REF;
+}
+
+std::map<Symbol*, RefInfo*> infoMap;
+typedef std::map<Symbol*, RefInfo*>::iterator RefInfoIter;
+
+static bool inferConstRef(Symbol*);
+
+static bool handleRHS(CallExpr* parent, CallExpr* call, SymExpr* se) {
+  SymExpr* LHS = toSymExpr(parent->get(1));
+  switch (call->primitive->tag) {
+    case PRIM_GET_MEMBER:
+    case PRIM_GET_MEMBER_VALUE:
+    case PRIM_GET_SVEC_MEMBER:
+    case PRIM_GET_SVEC_MEMBER_VALUE:
+    case PRIM_GET_REAL:
+    case PRIM_GET_IMAG:
+    case PRIM_ADDR_OF:
+    case PRIM_SET_REFERENCE: {
+      if (LHS->isRef()) {
+        if (!inferConstRef(LHS->var)) {
+          return false;
+        }
+      }
+    }
+
+    case PRIM_NOOP:
+    case PRIM_UNARY_MINUS:
+    case PRIM_UNARY_PLUS:
+    case PRIM_UNARY_NOT:
+    case PRIM_UNARY_LNOT:
+    case PRIM_ADD:
+    case PRIM_SUBTRACT:
+    case PRIM_MULT:
+    case PRIM_DIV:
+    case PRIM_MOD:
+    case PRIM_LSH:
+    case PRIM_RSH:
+    case PRIM_EQUAL:
+    case PRIM_NOTEQUAL:
+    case PRIM_LESSOREQUAL:
+    case PRIM_GREATEROREQUAL:
+    case PRIM_LESS:
+    case PRIM_GREATER:
+    case PRIM_AND:
+    case PRIM_OR:
+    case PRIM_XOR:
+    case PRIM_POW:
+    case PRIM_MIN:
+    case PRIM_MAX:
+
+    case PRIM_CHECK_NIL:
+    case PRIM_LOCAL_CHECK:
+    case PRIM_PTR_EQUAL:
+    case PRIM_PTR_NOTEQUAL:
+
+    case PRIM_BLOCK_LOCAL:
+    case PRIM_ON_LOCALE_NUM:
+
+    case PRIM_GET_SERIAL:
+    case PRIM_SET_SERIAL:
+
+    case PRIM_START_RMEM_FENCE:
+    case PRIM_FINISH_RMEM_FENCE:
+
+    case PRIM_SIZEOF:
+
+    case PRIM_GET_USER_LINE:
+    case PRIM_GET_USER_FILE:
+    case PRIM_LOOKUP_FILENAME:
+    case PRIM_WIDE_GET_LOCALE:
+    case PRIM_WIDE_GET_NODE:
+
+    // Just a 'read'
+    case PRIM_DEREF:
+
+    case PRIM_TYPEOF:
+    case PRIM_TYPE_TO_STRING:
+    case PRIM_ENUM_MIN_BITS:
+    case PRIM_ENUM_IS_SIGNED:
+    case PRIM_IS_UNION_TYPE:
+    case PRIM_IS_ATOMIC_TYPE:
+    case PRIM_IS_TUPLE_TYPE:
+    case PRIM_IS_STAR_TUPLE_TYPE:
+    case PRIM_IS_SUBTYPE:
+    case PRIM_IS_WIDE_PTR:
+    case PRIM_ERROR:
+    case PRIM_WARNING:
+
+    case PRIM_BLOCK_PARAM_LOOP:
+    case PRIM_BLOCK_BEGIN:
+    case PRIM_BLOCK_COBEGIN:
+    case PRIM_BLOCK_COFORALL:
+    case PRIM_BLOCK_ON:
+    case PRIM_BLOCK_BEGIN_ON:
+    case PRIM_BLOCK_COBEGIN_ON:
+    case PRIM_BLOCK_COFORALL_ON:
+    case PRIM_BLOCK_UNLOCAL:
+      return true;
+
+    // Unsure how to handle these yet, so to be safe we'll just return false.
```

Comment:
My intention was to force developers to think about this case when adding a new primitive, as I think the compiler should warn of one of the cases is not handled.

I'd be fine with going to a default-false statement, since they're equally conservative.